First will start by importing the libraries that we need Pandas, Numpy, matplotlib & seaborn

In [35]:
import pandas as pd
from pandas import DataFrame, Series
import numpy as np
import matplotlib.pyplot as plt
from pylab import rcParams
import seaborn as sb

Then we need to read our data

In [36]:
df = pd.read_csv("C:\\Users\\mrnaz\\Desktop\\BootCamp\\My_Project\\WN-JAN-2020.csv")

let's have a look at what just we read

In [37]:
df.head(5)

DAY_OF_MONTH  DAY_OF_WEEK   FL_DATE  MKT_CARRIER_FL_NUM TAIL_NUM ORIGIN  \
0             1            3  1/1/2020                5888   N951WN    ONT   
1             1            3  1/1/2020                6276   N467WN    ONT   
2             1            3  1/1/2020                4598   N7885A    ONT   
3             1            3  1/1/2020                4761   N551WN    ONT   
4             1            3  1/1/2020                5162   N968WN    ONT   

  ORIGIN_CITY_NAME ORIGIN_STATE_NM DEST     DEST_CITY_NAME  ...  \
0      Ontario, CA      California  SFO  San Francisco, CA  ...   
1      Ontario, CA      California  SFO  San Francisco, CA  ...   
2      Ontario, CA      California  SJC       San Jose, CA  ...   
3      Ontario, CA      California  SJC       San Jose, CA  ...   
4      Ontario, CA      California  SJC       San Jose, CA  ...   

  CANCELLATION_CODE  CRS_ELAPSED_TIME  ACTUAL_ELAPSED_TIME  AIR_TIME  \
0               NaN                95                122.0      74.0   
1               NaN                90                 92.0      71.0   
2               NaN                70                 68.0      57.0   
3               NaN                75                 75.0      63.0   
4               NaN                80                 67.0      57.0   

   DISTANCE CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
0       363           8.0            0.0       27.0             0.0   
1       363           NaN            NaN        NaN             NaN   
2       333           NaN            NaN        NaN             NaN   
3       333           NaN            NaN        NaN             NaN   
4       333           NaN            NaN        NaN             NaN   

   LATE_AIRCRAFT_DELAY  
0                 33.0  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 36 columns]

First I will rename some columns 

In [38]:
df.rename({'MKT_CARRIER_FL_NUM':'FL_NUM', 
               'ORIGIN_CITY_NAME':'ORIGIN_CITY',
               'ORIGIN_CITY_NAME':'ORIGIN_CITY',
               'ORIGIN_STATE_NM':'ORIGIN_STATE',
               'DEST_CITY_NAME':'DEST_CITY',
              'DAY_OF_MONTH':'MONTH_DAY', 
               'DAY_OF_WEEK':'WEEK_DAY',
               'DEST_STATE_NM':'DEST_STATE'},
              axis=1, inplace=True)

Second I will remove the Time in Hourly Block and the 3 letters city code it is not familiar for most people that are not working in the aviation let's just remove it and save some space

In [39]:
df.drop(['ORIGIN','DEST','DEP_TIME_BLK','ARR_TIME_BLK'],axis=1,inplace=True)

let's have a look on the table after we do this changes

In [40]:
df.head(5)  # or df.sample(5)

MONTH_DAY  WEEK_DAY   FL_DATE  FL_NUM TAIL_NUM  ORIGIN_CITY ORIGIN_STATE  \
0          1         3  1/1/2020    5888   N951WN  Ontario, CA   California   
1          1         3  1/1/2020    6276   N467WN  Ontario, CA   California   
2          1         3  1/1/2020    4598   N7885A  Ontario, CA   California   
3          1         3  1/1/2020    4761   N551WN  Ontario, CA   California   
4          1         3  1/1/2020    5162   N968WN  Ontario, CA   California   

           DEST_CITY  DEST_STATE  CRS_DEP_TIME  ...  CANCELLATION_CODE  \
0  San Francisco, CA  California          1810  ...                NaN   
1  San Francisco, CA  California          1150  ...                NaN   
2       San Jose, CA  California          2020  ...                NaN   
3       San Jose, CA  California          1340  ...                NaN   
4       San Jose, CA  California           915  ...                NaN   

   CRS_ELAPSED_TIME  ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  CARRIER_DELAY  \
0                95                122.0      74.0       363            8.0   
1                90                 92.0      71.0       363            NaN   
2                70                 68.0      57.0       333            NaN   
3                75                 75.0      63.0       333            NaN   
4                80                 67.0      57.0       333            NaN   

   WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0            0.0       27.0             0.0                 33.0  
1            NaN        NaN             NaN                  NaN  
2            NaN        NaN             NaN                  NaN  
3            NaN        NaN             NaN                  NaN  
4            NaN        NaN             NaN                  NaN  

[5 rows x 32 columns]

O.K we need to know more about our table such as number of non-null and data type for each column

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 109770 entries, 0 to 109769
Data columns (total 32 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   MONTH_DAY            109770 non-null  int64  
 1   WEEK_DAY             109770 non-null  int64  
 2   FL_DATE              109770 non-null  object 
 3   FL_NUM               109770 non-null  int64  
 4   TAIL_NUM             109674 non-null  object 
 5   ORIGIN_CITY          109770 non-null  object 
 6   ORIGIN_STATE         109770 non-null  object 
 7   DEST_CITY            109770 non-null  object 
 8   DEST_STATE           109770 non-null  object 
 9   CRS_DEP_TIME         109770 non-null  int64  
 10  DEP_TIME             107879 non-null  float64
 11  DEP_DELAY            107879 non-null  float64
 12  DEP_DELAY_NEW        107879 non-null  float64
 13  TAXI_OUT             107860 non-null  float64
 14  WHEELS_OFF           107860 non-null  float64
 15  WHEELS_ON        

In [42]:
df.isna().any()

MONTH_DAY              False
WEEK_DAY               False
FL_DATE                False
FL_NUM                 False
TAIL_NUM                True
ORIGIN_CITY            False
ORIGIN_STATE           False
DEST_CITY              False
DEST_STATE             False
CRS_DEP_TIME           False
DEP_TIME                True
DEP_DELAY               True
DEP_DELAY_NEW           True
TAXI_OUT                True
WHEELS_OFF              True
WHEELS_ON               True
TAXI_IN                 True
CRS_ARR_TIME           False
ARR_TIME                True
ARR_DELAY               True
ARR_DELAY_NEW           True
CANCELLED              False
CANCELLATION_CODE       True
CRS_ELAPSED_TIME       False
ACTUAL_ELAPSED_TIME     True
AIR_TIME                True
DISTANCE               False
CARRIER_DELAY           True
WEATHER_DELAY           True
NAS_DELAY               True
SECURITY_DELAY          True
LATE_AIRCRAFT_DELAY     True
dtype: bool

we need to know number of NaN in each colomn

In [43]:
df.isna().sum()  # or df.isnull().sum()

MONTH_DAY                   0
WEEK_DAY                    0
FL_DATE                     0
FL_NUM                      0
TAIL_NUM                   96
ORIGIN_CITY                 0
ORIGIN_STATE                0
DEST_CITY                   0
DEST_STATE                  0
CRS_DEP_TIME                0
DEP_TIME                 1891
DEP_DELAY                1891
DEP_DELAY_NEW            1891
TAXI_OUT                 1910
WHEELS_OFF               1910
WHEELS_ON                1956
TAXI_IN                  1956
CRS_ARR_TIME                0
ARR_TIME                 1956
ARR_DELAY                2062
ARR_DELAY_NEW            2062
CANCELLED                   0
CANCELLATION_CODE      107849
CRS_ELAPSED_TIME            0
ACTUAL_ELAPSED_TIME      2062
AIR_TIME                 2062
DISTANCE                    0
CARRIER_DELAY           99449
WEATHER_DELAY           99449
NAS_DELAY               99449
SECURITY_DELAY          99449
LATE_AIRCRAFT_DELAY     99449
dtype: int64

In [8]:
df.describe()

DAY_OF_MONTH    DAY_OF_WEEK  MKT_CARRIER_FL_NUM   CRS_DEP_TIME  \
count  109770.000000  109770.000000       109770.000000  109770.000000   
mean       15.996948       3.935602         2243.172251    1337.307780   
std         8.994348       1.908825         1705.824091     478.999221   
min         1.000000       1.000000            1.000000     500.000000   
25%         8.000000       2.000000          966.000000     915.000000   
50%        16.000000       4.000000         1744.000000    1330.000000   
75%        24.000000       5.000000         3241.750000    1740.000000   
max        31.000000       7.000000         6860.000000    2335.000000   

            DEP_TIME      DEP_DELAY  DEP_DELAY_NEW       TAXI_OUT  \
count  107879.000000  107879.000000  107879.000000  107860.000000   
mean     1341.388871       4.342968       6.615319      12.020842   
std       485.342522      21.106205      20.222221       5.975671   
min         1.000000     -20.000000       0.000000       1.000000   
25%       919.000000      -4.000000       0.000000       9.000000   
50%      1335.000000      -1.000000       0.000000      10.000000   
75%      1746.000000       4.000000       4.000000      14.000000   
max      2359.000000     602.000000     602.000000     115.000000   

          WHEELS_OFF      WHEELS_ON  ...      CANCELLED  CRS_ELAPSED_TIME  \
count  107860.000000  107814.000000  ...  109770.000000     109770.000000   
mean     1361.662618    1482.372326  ...       0.017500        128.869591   
std       484.029902     516.422930  ...       0.131126         56.278200   
min         1.000000       1.000000  ...       0.000000         35.000000   
25%       931.000000    1059.000000  ...       0.000000         85.000000   
50%      1347.000000    1515.000000  ...       0.000000        115.000000   
75%      1758.000000    1918.000000  ...       0.000000        160.000000   
max      2400.000000    2400.000000  ...       1.000000        385.000000   

       ACTUAL_ELAPSED_TIME       AIR_TIME       DISTANCE  CARRIER_DELAY  \
count        107708.000000  107708.000000  109770.000000   10321.000000   
mean            119.095796     102.059011     724.337952      18.209185   
std              55.411353      54.368042     441.776658      32.635017   
min              27.000000      18.000000      84.000000       0.000000   
25%              76.000000      60.000000     371.000000       0.000000   
50%             106.000000      89.000000     616.000000       7.000000   
75%             149.000000     131.000000     971.000000      21.000000   
max             416.000000     389.000000    2557.000000     422.000000   

       WEATHER_DELAY     NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
count   10321.000000  10321.000000    10321.000000         10321.000000  
mean        0.838485      7.873946        0.097568            18.690631  
std        10.117391     18.111303        2.034031            34.640534  
min         0.000000      0.000000        0.000000             0.000000  
25%         0.000000      0.000000        0.000000             0.000000  
50%         0.000000      0.000000        0.000000             4.000000  
75%         0.000000     10.000000        0.000000            24.000000  
max       407.000000    280.000000      117.000000           597.000000  

[8 rows x 25 columns]

To examine data types and look for cases of missing or potentially wrong data, the describe function helps to get an overview of the data from a statistical summary point of view. This is also useful for spotting any potential errors that may need a closer look.

In [44]:
df.shape

(109770, 32)

There are 109770 records and 36 columns in the dataset.

In [45]:
df.size

3512640

In [46]:
df.columns 

Index(['MONTH_DAY', 'WEEK_DAY', 'FL_DATE', 'FL_NUM', 'TAIL_NUM', 'ORIGIN_CITY',
       'ORIGIN_STATE', 'DEST_CITY', 'DEST_STATE', 'CRS_DEP_TIME', 'DEP_TIME',
       'DEP_DELAY', 'DEP_DELAY_NEW', 'TAXI_OUT', 'WHEELS_OFF', 'WHEELS_ON',
       'TAXI_IN', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY', 'ARR_DELAY_NEW',
       'CANCELLED', 'CANCELLATION_CODE', 'CRS_ELAPSED_TIME',
       'ACTUAL_ELAPSED_TIME', 'AIR_TIME', 'DISTANCE', 'CARRIER_DELAY',
       'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY'],
      dtype='object')

In [47]:
df.dtypes # This will peovied data type for each column

MONTH_DAY                int64
WEEK_DAY                 int64
FL_DATE                 object
FL_NUM                   int64
TAIL_NUM                object
ORIGIN_CITY             object
ORIGIN_STATE            object
DEST_CITY               object
DEST_STATE              object
CRS_DEP_TIME             int64
DEP_TIME               float64
DEP_DELAY              float64
DEP_DELAY_NEW          float64
TAXI_OUT               float64
WHEELS_OFF             float64
WHEELS_ON              float64
TAXI_IN                float64
CRS_ARR_TIME             int64
ARR_TIME               float64
ARR_DELAY              float64
ARR_DELAY_NEW          float64
CANCELLED                int64
CANCELLATION_CODE       object
CRS_ELAPSED_TIME         int64
ACTUAL_ELAPSED_TIME    float64
AIR_TIME               float64
DISTANCE                 int64
CARRIER_DELAY          float64
WEATHER_DELAY          float64
NAS_DELAY              float64
SECURITY_DELAY         float64
LATE_AIRCRAFT_DELAY    float64
dtype: o

In [49]:
# i need to list all origin city name
print(df.ORIGIN_CITY.unique())

['Ontario, CA' 'Norfolk, VA' 'West Palm Beach/Palm Beach, FL'
 'Portland, OR' 'Philadelphia, PA' 'Phoenix, AZ' 'Pittsburgh, PA'
 'Pensacola, FL' 'Providence, RI' 'Portland, ME' 'Raleigh/Durham, NC'
 'Richmond, VA' 'Reno, NV' 'Rochester, NY' 'Fort Myers, FL'
 'San Diego, CA' 'San Antonio, TX' 'Louisville, KY' 'Seattle, WA'
 'San Francisco, CA' 'San Jose, CA' 'San Juan, PR' 'Salt Lake City, UT'
 'Sacramento, CA' 'Santa Ana, CA' 'St. Louis, MO' 'Tampa, FL' 'Tulsa, OK'
 'Tucson, AZ' 'Albuquerque, NM' 'Albany, NY' 'Amarillo, TX' 'Atlanta, GA'
 'Austin, TX' 'Dallas, TX' 'New Orleans, LA' 'Oakland, CA' 'Kahului, HI'
 'Oklahoma City, OK' 'Omaha, NE' 'Hartford, CT' 'Birmingham, AL'
 'Nashville, TN' 'Boise, ID' 'Boston, MA' 'Buffalo, NY' 'Burbank, CA'
 'Baltimore, MD' 'Charleston, SC' 'Cleveland, OH' 'Charlotte, NC'
 'Columbus, OH' 'Corpus Christi, TX' 'Cincinnati, OH' 'Washington, DC'
 'Denver, CO' 'Des Moines, IA' 'Detroit, MI' 'Panama City, FL'
 'El Paso, TX' 'Fort Lauderdale, FL' 'Spokane, W

In [50]:
#To check if there is duplicate rows in the dataset
sum(df.duplicated())

0

In [51]:
# This command tell us how many flights in each airport, it will basically calculate city name repeat number
print (df['ORIGIN_CITY'].value_counts())

Dallas, TX         5840
Las Vegas, NV      5727
Chicago, IL        5721
Denver, CO         5636
Baltimore, MD      5457
                   ... 
Panama City, FL      98
Pensacola, FL        93
Portland, ME         91
Des Moines, IA       89
Hilo, HI             51
Name: ORIGIN_CITY, Length: 88, dtype: int64
